src/main/java/bio/terra/service/filedata/DrsService.java

```diff
@@ -52,17 +64,46 @@ public DrsService(SnapshotService snapshotService,
                       DrsIdService drsIdService,
                       IamService samService,
                       ResourceService resourceService,
+                      ConfigurationService configurationService,
+                      KubeService kubeService,
                       PerformanceLogger performanceLogger) {
         this.snapshotService = snapshotService;
         this.fileService = fileService;
         this.drsIdService = drsIdService;
         this.samService = samService;
         this.resourceService = resourceService;
+        this.configurationService = configurationService;
+        this.kubeService = kubeService;
         this.performanceLogger = performanceLogger;
     }
 
-    public DRSObject lookupObjectByDrsId(AuthenticatedUserRequest authUser, String drsObjectId, Boolean expand) {
+    public void decrement() {
+        int currentValue = currentDRSRequests.get();
+        if (currentValue > 0) {
+            int newValue = currentValue - 1;
+            currentDRSRequests.set(newValue);
+        }
+    }
+
+    public void increment() {
+        int currentValue = currentDRSRequests.get();
+        int newValue = currentValue + 1;
+        currentDRSRequests.set(newValue);
+    }
+
+    public void checkLookupMax() throws TooManyRequestsException {
```

Comment:
Since `increment()` implies that a call will be made, I would do the max check there instead of making it a separate method. My reading of this pattern is:

- acquire resource
  - do operation
- release resource

The way I think about max is, you can' t acquire the resource if there aren't any free. So the acquire step fails and the operation isn't started.

Reply:
I realize this comment may be too vague to do anything with so here's what I have in mind. Once this case is identified as a resource acquisition problem, there are a few patterns you can use to approach it. I would use RAII. My preferred way in java is to use `try-with-resources`:

```
    private class DrsRequestResource implements AutoCloseable {
        DrsRequestResource() {
            // make sure not too many requests are being made at once
            int podCount = kubeService.getActivePodCount();
            int maxDRSLookups = configurationService.getParameterValue(ConfigEnum.DRS_LOOKUP_MAX);
            int max = maxDRSLookups / podCount;
            if (currentDRSRequests.incrementAndGet() > max) {
                throw new TooManyRequestsException("Too many requests are being made at once. Please try again later.");
            }
        }

        @Override
        public void close() {
            currentDRSRequests.decrementAndGet();
        }
    }

    public DRSObject lookupObjectByDrsId(AuthenticatedUserRequest authUser, String drsObjectId, Boolean expand
    ) throws TooManyRequestsException {
        try (DrsRequestResource r = new DrsRequestResource()) {
            DrsId drsId = drsIdService.fromObjectId(drsObjectId);
            SnapshotProject snapshotProject = null;
            [...]

            throw new IllegalArgumentException("Invalid object type");
        }
    }
```
Since try-with-resources will close `r` at the end of the block, even if an exception occurs, the resource will always be released.